plugins/decklink/decklink-output.cpp

```diff
@@ -0,0 +1,237 @@
+{
+	auto *decklink = (DeckLinkOutput *)data;
+	delete decklink;
+}
+
+static void *decklink_output_create(obs_data_t *settings, obs_output_t *output)
+{
+	auto *decklinkOutput = new DeckLinkOutput(output, deviceEnum);
+
+	decklinkOutput->deviceHash = obs_data_get_string(settings, DEVICE_HASH);
+	decklinkOutput->modeID = obs_data_get_int(settings, MODE_ID);
+
+	return decklinkOutput;
+}
+
+static void decklink_output_update(void *data, obs_data_t *settings)
+{
+	auto *decklink = (DeckLinkOutput *)data;
+
+	decklink->deviceHash = obs_data_get_string(settings, DEVICE_HASH);
+	decklink->modeID = obs_data_get_int(settings, MODE_ID);
+}
+
+static bool decklink_output_start(void *data)
+{
+	auto *decklink = (DeckLinkOutput *)data;
+	struct obs_audio_info aoi;
+
+	if (!obs_get_audio_info(&aoi)) {
+		blog(LOG_WARNING, "No active audio");
+		return false;
+	}
+
+	decklink->audio_samplerate = aoi.samples_per_sec;
+	decklink->audio_planes = 2;
+	decklink->audio_size = get_audio_size(AUDIO_FORMAT_16BIT, aoi.speakers, 1);
+
+	decklink->start_timestamp = 0;
+
+	ComPtr<DeckLinkDevice> device;
+
+	device.Set(deviceEnum->FindByHash(decklink->deviceHash));
+
+	DeckLinkDeviceMode *mode = device->FindOutputMode(decklink->modeID);
+
+	decklink->SetSize(mode->GetWidth(), mode->GetHeight());
+
+	struct video_scale_info to = {};
+	to.format = VIDEO_FORMAT_UYVY;
+	to.width = mode->GetWidth();
+	to.height =  mode->GetHeight();
+
+	obs_output_set_video_conversion(decklink->GetOutput(), &to);
+
+	decklink->Activate(device, decklink->modeID);
+
+	audio_convert_info conversion{};
+	conversion.format = AUDIO_FORMAT_16BIT;
+	conversion.speakers = SPEAKERS_STEREO;
+
+	obs_output_set_audio_conversion(decklink->GetOutput(), &conversion);
+
+	if (!obs_output_begin_data_capture(decklink->GetOutput(), 0))
+		return false;
+
+	return true;
+}
+
+static void decklink_output_stop(void *data, uint64_t ts)
+{
+	auto *decklink = (DeckLinkOutput *)data;
+
+	obs_output_end_data_capture(decklink->GetOutput());
+
+	ComPtr<DeckLinkDevice> device;
+
+	device.Set(deviceEnum->FindByHash(decklink->deviceHash));
+
+	decklink->Deactivate();
+}
+
+static void decklink_output_raw_video(void *data, struct video_data *frame)
+{
+	auto *decklink = (DeckLinkOutput *)data;
+
+	if (!decklink->start_timestamp)
+		decklink->start_timestamp = frame->timestamp;
+
+	decklink->DisplayVideoFrame(frame);
+}
+
+static bool prepare_audio(DeckLinkOutput *decklink,
+		const struct audio_data *frame,
+		struct audio_data *output)
+{
+	*output = *frame;
+
+	if (frame->timestamp < decklink->start_timestamp) {
+		uint64_t duration = (uint64_t)frame->frames * 1000000000 /
+				(uint64_t)decklink->audio_samplerate;
+		uint64_t end_ts = (frame->timestamp + duration);
+		uint64_t cutoff;
+
+		if (end_ts <= decklink->start_timestamp)
+			return false;
+
+		cutoff = decklink->start_timestamp - frame->timestamp;
+		output->timestamp += cutoff;
+
+		cutoff = cutoff * (uint64_t)decklink->audio_samplerate /
```

Comment:
I think you can use `*=` here, since `*` and `/` have the same precedence and are evaluated left-to-right.  Like so:
```
		cutoff *= (uint64_t)decklink->audio_samplerate / 1000000000;
```